server/src/main/java/org/candlepin/liquibase/PerOrgProductsMigrationTask.java

```diff
@@ -0,0 +1,455 @@
+    private Timestamp migrationTime;
+
+    public PerOrgProductsMigrationTask(Database database, CustomTaskLogger logger) {
+        super(database, logger);
+
+        this.migratedProducts = new HashMap<String, String>();
+        this.migratedContent = new HashMap<String, String>();
+        this.migrationTime = new Timestamp(System.currentTimeMillis());
+    }
+
+    /**
+     * Generates a prepared statement for performing a bulk insert into the given table
+     *
+     * @param table
+     *  the table into which to insert the data
+     *
+     * @param rows
+     *  the number of rows being inserted
+     *
+     * @param cols
+     *  the columns receiving data for each row
+     *
+     * @return
+     *  a PreparedStatement instance representing the bulk insert operation
+     */
+    protected PreparedStatement generateBulkInsertStatement(String table, int rows, String... cols)
+        throws DatabaseException, SQLException {
+
+        if (rows > 0) {
+            StringBuilder builder, rowbuilder;
+
+            if (!this.database.getDatabaseProductName().matches(".*(?i:oracle).*")) {
+                rowbuilder = new StringBuilder(2 + cols.length << 1);
+
+                rowbuilder.append('(');
+                for (int i = 0; i < cols.length; ++i) {
+                    rowbuilder.append("?,");
+                }
+                rowbuilder.deleteCharAt(rowbuilder.length() - 1).append(')');
+
+                builder = new StringBuilder(15 + 20 * cols.length + table.length() +
+                    (2 + cols.length * 3) * rows);
+
+                builder.append("INSERT INTO ").append(table).append('(');
+                for (String column : cols) {
+                    builder.append(column).append(',');
+                }
+                builder.deleteCharAt(builder.length() - 1).append(") VALUES");
+
+                for (int i = 0; i < rows; ++i) {
+                    builder.append(rowbuilder).append(',');
+                }
+                builder.deleteCharAt(builder.length() - 1);
+            }
+            else {
+                rowbuilder = new StringBuilder(20 * cols.length + table.length());
+
+                rowbuilder.append("INTO ").append(table).append('(');
+                for (String column : cols) {
+                    rowbuilder.append(column).append(',');
+                }
+                rowbuilder.deleteCharAt(rowbuilder.length() - 1).append(") VALUES(");
+                for (int i = 0; i < cols.length; ++i) {
+                    rowbuilder.append("?,");
+                }
+                rowbuilder.deleteCharAt(rowbuilder.length() - 1).append(") ");
+
+                builder = new StringBuilder(30 + (20 * cols.length + table.length()) * rows);
+                builder.append("INSERT ALL ");
+
+                for (int i = 0; i < rows; ++i) {
+                    builder.append(rowbuilder);
+                }
+
+                builder.append("SELECT 1 FROM DUAL");
```

Comment:
Oracle shenanigans. The syntax requires that you do finish an INSERT ALL with a SELECT of some kind, and "SELECT 1" by itself isn't (always) valid there, so we use an internal table that should always exist.